pkg/cacheutil/redis_client.go

```diff
@@ -0,0 +1,316 @@
+	MaxConnAge time.Duration `yaml:"max_conn_age"`
+
+	// MaxGetMultiConcurrency specifies the maximum number of concurrent GetMulti() operations.
+	// If set to 0, concurrency is unlimited.
+	MaxGetMultiConcurrency int `yaml:"max_get_multi_concurrency"`
+
+	// GetMultiBatchSize specifies the maximum size per batch for mget.
+	GetMultiBatchSize int `yaml:"get_multi_batch_size"`
+
+	// MaxSetMultiConcurrency specifies the maximum number of concurrent SetMulti() operations.
+	// If set to 0, concurrency is unlimited.
+	MaxSetMultiConcurrency int `yaml:"max_set_multi_concurrency"`
+
+	// SetMultiBatchSize specifies the maximum size per batch for pipeline set.
+	SetMultiBatchSize int `yaml:"set_multi_batch_size"`
+}
+
+func (c *RedisClientConfig) validate() error {
+	if c.Addr == "" {
+		return errors.New("no redis addr provided")
+	}
+	return nil
+}
+
+// RedisClient is a wrap of redis.Client.
+type RedisClient struct {
+	*redis.Client
+	config RedisClientConfig
+
+	// getMultiGate used to enforce the max number of concurrent GetMulti() operations.
+	getMultiGate gate.Gate
+	// setMultiGate used to enforce the max number of concurrent SetMulti() operations.
+	setMultiGate gate.Gate
+
+	logger           log.Logger
+	durationSet      prometheus.Observer
+	durationSetMulti prometheus.Observer
+	durationGetMulti prometheus.Observer
+}
+
+// NewRedisClient makes a new RedisClient.
+func NewRedisClient(logger log.Logger, name string, conf []byte, reg prometheus.Registerer) (*RedisClient, error) {
+	config, err := parseRedisClientConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewRedisClientWithConfig(logger, name, config, reg)
+}
+
+// NewRedisClientWithConfig makes a new RedisClient.
+func NewRedisClientWithConfig(logger log.Logger, name string, config RedisClientConfig,
+	reg prometheus.Registerer) (*RedisClient, error) {
+
+	if err := config.validate(); err != nil {
+		return nil, err
+	}
+	redisClient := redis.NewClient(&redis.Options{
+		Addr:         config.Addr,
+		Username:     config.Username,
+		Password:     config.Password,
+		DB:           config.DB,
+		DialTimeout:  config.DialTimeout,
+		ReadTimeout:  config.ReadTimeout,
+		WriteTimeout: config.WriteTimeout,
+		MinIdleConns: config.MinIdleConns,
+		MaxConnAge:   config.MaxConnAge,
+		IdleTimeout:  config.IdleTimeout,
+	})
+
+	if reg != nil {
+		reg = prometheus.WrapRegistererWith(prometheus.Labels{"name": name}, reg)
+	}
+
+	c := &RedisClient{
+		Client: redisClient,
+		config: config,
+		logger: logger,
+		getMultiGate: gate.New(
+			extprom.WrapRegistererWithPrefix("thanos_redis_getmulti_", reg),
+			config.MaxGetMultiConcurrency,
+		),
+		setMultiGate: gate.New(
+			extprom.WrapRegistererWithPrefix("thanos_redis_setmulti_", reg),
+			config.MaxSetMultiConcurrency,
+		),
+	}
+	duration := promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "thanos_redis_operation_duration_seconds",
+		Help:    "Duration of operations against memcached.",
```

Comment:
Typo: should be _Duration of operations against **redis**._

Reply:
let me fix it.

Reply:
let me fix it.